### ISSUE TITLE
Login several users / agents simoultanesly

### ISSUE THREAD

**Issue by Zenkit** [visionmedia/supertest#287]
Hello,

I am using supertest / superagent to test my application. I would like to be able to log in several users at once to test different interactions. So far I havent been able to do that.

```
global.request = require('supertest');
global.TestAgent = request.agent;

agent = new TestAgent(express);
agent2 = new TestAgent(express);
agentAdmin = new TestAgent(express);
```

Each agent here, resembles the exact same user. Is there something I am missing?

Kind Regards

**Comment by AhmedAKZM** [comment]
What do you mean exactly when you say they resemble the "exact same user"? Are the cookie jars being shared? I need a bit more information to reproduce your errors.

In any case, I've been able to successfully create multiple agents in my tests.

**Comment by Zenkit** [comment]
yes the creation of multiple agents works here too.

I am login in agent as testuser, agent2 as testuser2, agentAdmin as adminuser

when calling my /api/auth/currentuser route with any of the agents, I always get the first one i logged in, instead of the one "belonging" to that agent.

```
Are the cookie jars being shared?
```
Might be something like that, I haven't debugged into the superagent /supertest code.

**Comment by AhmedAKZM** [comment]
I have a similar route in my application (/api/users/me), and I just tested it with three different agents (no authentication for one, and two different accounts for the others), and the responses I got are:

```json
// Response for unauthenticated agent 
{ "statusCode": 401,
  "error": "Unauthorized",
  "message": "Missing authentication" }

// Response for agent #1
{ "user":
   { "id": 1000,
     "email": "existent@admins.testuser",
     "displayName": "New Display Name",
     "Image": { "id": 1, "path": "random.jpg" } } }

// Response for agent #2
{ "user":
   { "id": 1001,
     "email": "jane@members.testuser",
     "displayName": "Jane Doe",
     "Image": { "id": 15, "path": "user8.jpg" } } }
```

which is as expected. I couldn't figure out how to log the contents of an agent's cookie jar, but I did check the cookies being sent with the request through the response object. It'd be useful if you could do the same and let me know what you find. You could try something along the lines of:

```javascript
// set up and log agent, agent2, and agentAdmin in, and then get the current user for each

agent.get('/api/auth/currentuser').end(function(err, res) {
  console.log('\n' + JSON.stringify({
    agent: 'agent',
    cookies: res.request.cookies
  }, null, 4));
});

agent2.get('/api/auth/currentuser').end(function(err, res) {
  console.log('\n' + JSON.stringify({
    agent: 'agent2',
    cookies: res.request.cookies
  }, null, 4));
});

agentAdmin.get('/api/auth/currentuser').end(function(err, res) {
  console.log('\n' + JSON.stringify({
    agent: 'agentAdmin',
    cookies: res.request.cookies
  }, null, 4));
});
```

It would also be useful if you can log the headers of the response received when authenticating agents to make sure that the server isn't setting the cookies erroneously. A simple ```console.log``` on the response headers should do the trick. Look for the ```set-cookie``` header:

```javascript
yourAgent.post('/path/to/auth/endpoint')
  .send({ auth parameters })
  .end(function(err, res) {
    console.log('\nAuthentication headers: ' + JSON.stringify({
      agent: 'yourAgent',
      headers: res.headers
    }, null, 4));
  });
```

When running code similar to the above for agent1 and agent2 in my case I get:

```json
// agent1
Authentication headers: {                                                                                                                                        
    "agent": "agent1",                                                                                                                                   
    "headers": {                          
        "... other headers ..."                                                                               
        "set-cookie": [                                                                                                                                  
            "singlebody-auth=Fe26.2**f1127711a1b0967fe8cf532bbf583f50f547d5965fcdbaf82bf26e86562d0a60*lcQklQ6g3UOpsQSh1Vy8OA*12HWU4e6o_JDs5PK4JmHfw**76d9
4d523615ad93eb88906d9f0e000f0dc2d12d161bebe20f4a1b05d30cf4ab*wAo9C8cIkOdDVhq6HWwr5pZPHKK_7z5jRRwJaBS36P4; HttpOnly; Path=/"                              
        ],                                    
        "... other headers ..."                                                                                                       
    }                                                                                                                                                    
}               

// agent2
Authentication headers: {
    "agent": "agent2",
    "headers": {
        "... other headers ..."
        "set-cookie": [
            "singlebody-auth=Fe26.2**56e6d03bf3c8480692d004c5169ce665db439494dec42d3041a9c013b3f9f38f*DArn1Zk8mXC-dNBdL49mPg*lJMijDhDhJGfisYCAhDcLA**369a358d25918c9ff7b6aae5955a21cf721be0d05ceba9780b28c3a972c10552*dPtiDb2P4WoxscEe9j5fm9ESZRTa5NAtB0E5A1y4rgI; HttpOnly; Path=/"
        ],
        "... other headers ..."
    }
}                                                                                                                   
```

I suggest you check the cookies received while authenticating each agent, and then check the cookies sent to the ```currentuser``` endpoint. If they're not the same, then there's an issue with how you've set up the agents (see comment below), or an issue somewhere with how superagent/supertest is handling cookies on your end (working fine for me though). Let me know what you find.

Are you using the same instance of express for each agent? I'm not sure how superagent/supertest uses the ```app``` parameter provided during initialization exactly, but having each agent instantiated with the same express instance might be causing an issue. It might be worth trying to instantiate each agent with a different instance (provide a factory method in your main application module to create express instances or something), or try instead instantiating each agent with a URL (```require('supertest').agent('http://localhost:8000')```), and then having your express server run as another process while testing. I used the second method to instantiate agents (I think it causes an express instance to be created internally? I'm not familiar with the inner workings of superagent/supertest).

**Comment by Zenkit** [comment]
Thanks for all the help @AhmedAKZM.

I tried your suggestions and everything seemed to be fine, so I changed the express instance to the url and suddenly the tests started working. So I suspected the express instance to be the culprit and after getting my tests to work properly, changed it back to be sure. Suddenly it also worked with the express instance.

Sadly I cannot trace back where the error originated from, but I can confirm now, that it is working.

Thanks and Kind Regards,
Matthias Weidemann
